stp_zmq/test/test_stashed_client_messages.py

```diff
@@ -0,0 +1,190 @@
+    msg2 = {'msg': 'msg2'}
+
+    beta.send(msg1, alpha.listener.IDENTITY)
+    assert pending_client_messages[alpha.listener.IDENTITY] == [(0, msg1)]
+
+    alpha.connect(name=beta.name, ha=beta.ha,
+                  verKeyRaw=beta.verKeyRaw, publicKeyRaw=beta.publicKeyRaw)
+
+    looper.runFor(0.25)
+
+    alpha.send(msg2, beta.name)
+    looper.run(eventually(
+        lambda msg_handler: assertExp(msg_handler.received_messages == [msg1, msg2]),
+        alpha_handler))
+    assert not pending_client_messages
+
+
+def test_invalid_msgs_are_not_stashed(tdir, looper, stacks, alpha_handler, tconf):
+    alpha, beta = stacks
+    pending_client_messages = beta._client_message_provider._pending_client_messages
+    msg = {'msg': 'msg1' * tconf.MSG_LEN_LIMIT}
+    assert not pending_client_messages
+    assert not alpha_handler.received_messages
+
+    alpha.connect(name=beta.name, ha=beta.ha,
+                  verKeyRaw=beta.verKeyRaw, publicKeyRaw=beta.publicKeyRaw)
+
+    looper.runFor(0.25)
+
+    alpha.send(msg, beta.name)
+    assert not pending_client_messages
+    assert not alpha_handler.received_messages
+
+    msg = {'msg': 'msg1'}
+    alpha.send(msg, beta.name)
+    looper.run(eventually(
+        lambda msg_handler: assertExp(msg_handler.received_messages == [msg]),
+        alpha_handler))
+
+
+def test_resending_for_old_stash_msgs(tdir, tconf, looper, stacks,
+                                      alpha_handler, monkeypatch):
+    alpha, beta = stacks
+    msg1 = {'msg': 'msg1'}
+    pending_client_messages = beta._client_message_provider._pending_client_messages
+
+    alpha.connect(name=beta.name, ha=beta.ha,
+                  verKeyRaw=beta.verKeyRaw, publicKeyRaw=beta.publicKeyRaw)
+    looper.runFor(0.25)
+
+    def fake_send_multipart(msg_parts, flags=0, copy=True, track=False, **kwargs):
+        raise zmq.Again
+
+    monkeypatch.setattr(beta.listener, 'send_multipart',
+                        fake_send_multipart)
+    alpha.send(msg1, beta.name)
+    looper.run(eventually(
+        lambda messages: assertExp(messages[alpha.listener.IDENTITY] == [(0, msg1)]),
+        pending_client_messages))
+    monkeypatch.undo()
+
+    beta._client_message_provider._timer.set_time(tconf.RESEND_CLIENT_MSG_TIMEOUT + 2)
+    looper.run(eventually(
+        lambda msg_handler: assertExp(msg_handler.received_messages == [msg1]),
+        alpha_handler))
+    assert not pending_client_messages
+
+
+def test_limit_msgs_for_client(tconf, looper, stacks, alpha_handler):
+    alpha, beta = stacks
+    pending_client_messages = beta._client_message_provider._pending_client_messages
+    for i in range(tconf.PENDING_MESSAGES_FOR_ONE_CLIENT_LIMIT + 2):
+        beta.send(create_msg(i),
+                  alpha.listener.IDENTITY)
+    assert len(pending_client_messages[alpha.listener.IDENTITY]) == tconf.PENDING_MESSAGES_FOR_ONE_CLIENT_LIMIT
+    assert pending_client_messages[alpha.listener.IDENTITY][0] != (0, create_msg(0))
```

Comment:
Can we check what is `pending_client_messages[alpha.listener.IDENTITY][0]` value? This is `(0, create_msg(2)`, isn't it?